lib/stages/BeamBufferSort.cpp

```diff
@@ -0,0 +1,414 @@
+                            fill_freq_meta(sub_frame_metadata, ch_id, fpga_seq_start_at_dump, ctime_at_dump, dump_frame_metadata -> stream_id.id,\
+                                           dump_frame_metadata -> dataset_id, dump_frame_metadata -> beam_number, dump_frame_metadata -> ra, \
+				           dump_frame_metadata -> dec, dump_frame_metadata -> scaling);
+			}
+			
+                        // Copy data to out buffer
+			src_start = FreqIDBeamMeta_size + in_frame_dump_offset * num_pol;
+                        memcpy(&output[dest_start + FreqIDBeamMeta_size + output_buff_offset * num_pol], \
+			       &sort_queue[0][ch_id][src_start], copy_size * num_pol);	
+	            }
+                    
+		}
+		//INFO("Old offset in {:d} out {:d} copy {:d}", in_frame_dump_offset, output_buff_offset, copy_size);
+		in_frame_dump_offset += copy_size;
+		output_buff_offset += copy_size;
+		//INFO("Updated out offset {:d}", output_buff_offset);
+
+	    }
+	    // Update time offset
+	    time_offset = frame_nr - frame0;
+	}
+            
+	// Check if the time offset too old
+	if (time_offset < 0){ // The incoming frame is too old
+            INFO("Drop offset {:d} frame_nr{:d}, frame0 {:d}, freq ID {:d}", time_offset,
+                 frame_nr, frame0, freq_bin);
+            INFO("Frame {:d} comes too late and will be droped.", frame_nr);
+            mark_frame_empty(in_buf, unique_name.c_str(), input_frame_id);
+            input_frame_id++;
+            continue;
+        }
+	// Copy in frame to the queue.
+        //INFO("In queue {:d} {:d}", time_offset, freq_bin);
+	FreqIDBeamMetadata* sub_frame_metadata = (FreqIDBeamMetadata*)&sort_queue[time_offset][freq_bin][0];
+	memcpy(sub_frame_metadata, in_metadata, FreqIDBeamMeta_size);
+	// copy data
+	uint8_t* sub_frame_data = &sort_queue[time_offset][freq_bin][FreqIDBeamMeta_size];
+	memcpy(sub_frame_data, input, in_buf->frame_size);
+        // Debug line
+	FreqIDBeamMetadata* debug_meta = (FreqIDBeamMetadata*)&sort_queue[time_offset][freq_bin][0];
+        INFO("Debug RA {:f} Dec: {:f}, scaling: {:d}, beam_num: {:d}, freq_id {:d}, seq_start {:d} stream_id {:d} c_time {:d}\n", 
+             debug_meta->ra, debug_meta->dec, debug_meta->scaling,
+             debug_meta->beam_number, debug_meta->frequency_bin, debug_meta->fpga_seq_start, 
+             debug_meta->stream_id.id, debug_meta -> ctime.tv_sec);	
+
+	// mark fill status is done
+	queue_status[time_offset][freq_bin] = 1;
+	// Mark in frame empty
+	mark_frame_empty(in_buf, unique_name.c_str(), input_frame_id);
+        input_frame_id++;
+    }
+    // Free memory
+    // Free queue
+    for (int i = 0; i < wait_nframes; ++i){
+        for (uint32_t j = 0; j < total_freq_chan; j++){
+            free(sort_queue[i][j]);
+        }
+    // Free temp freqid metadata
+    free(temp_freq_metadata);
+    }
```

Comment:
Can this not be done in the destructor?

Reply:
temp_freq_metadata is defined in the `main_thread()`. I don't think it can be done in the destructor.